File: gene_network_modeling/transcription_ssa.py
import numpy as np
import matplotlib
matplotlib.use('tkagg')
import matplotlib.pyplot as plt
import random

X = [0]
t = [0]

tend = 1000

k = 2
gamma = 0.1

while t[-1] < tend:

        current_X = X[-1]

        rates = [k, gamma * current_X]
        rate_sum = sum(rates)

        tau = np.random.exponential(scale=1/rate_sum)

        t.append(t[-1] + tau)

        rand = random.uniform(0,1)

        # production event
        if rand * rate_sum > 0 and rand * rate_sum <= rates[0]:
                X.append(X[-1] + 1)

        # decay event
        elif rand * rate_sum > rates[0] and rand * rate_sum <= rates[0] + rates[1]:
                X.append(X[-1] - 1)



plt.plot(t,X)
plt.xlabel("time")
plt.ylabel("mRNA quantity")
plt.show()

File: gene_network_modeling/transcription_ssa.m


close all
clear all
clc


tend = 100;

k = 2;
gamma = 0.1;


t(1) = 0;
X(1) = 0;
i = 1;


while t(end)< tend
   
    rates = zeros(1,2);
    rates(1) = k;
    rates(2) = gamma * X(i);
    
    rate_sum = sum(rates);
    
    
    r1 = rand(1);
    tau = (1/rate_sum)*log(1/r1);
    
    t(i+1) = t(i) + tau;
    
    
    r2 = rand(1);
    
    if r2*rate_sum <= rates(1)
        X(i+1) = X(i) + 1;
        
    else if r2 * rate_sum > rates(1) && r2 * rate_sum <= rates(1) + rates(2)
            X(i+1) = X(i) - 1;
        end 
    end
    
    
    
    i = i + 1;
    
    
end


plot(t,X)

